### PR TITLE
Safer handling of the initial installation of add-ons

### LINKED ISSUE #16704: Add-on bundle installer: add-on is removed without installation if add-on installTasks.onInstall function fails on update
Hi,

A more general version of #16702:

### Steps to reproduce:
Create an add-on (or modify an existing add-on from its source code) with the following installTasks.py module (it works better if installTasks.py already exists in the source code):

`# Whatever copyright header you wish to use`

`def onInstall() -> None:`
`    raise RuntimeError(whateverExceptionText)`

Package the modified add-on, then try installing it. To fully reproduce the problem, you must create an add-on that is newer than an installed add-on (works best if you are add-on authors and ask an add-on to "volunteer" i.e. edit one of your add-ons with modified installTasks.py and reverting it afterwards). For a demo, I have modified a dev build of Windows App Essentials.

### Actual behavior:
After installation prompts are shown, the progress tones are heard (if NVDA was told to do so) along with error tone (if alpha build is in use), then NVDA asks for a restart.

previous add-on version is removed along with json files, without the new add-on being installed

### Expected behavior:
NVDA presents add-on installation error dialog.

### NVDA logs, crash dumps and other attachments:
ERROR - addonHandler.installAddonBundle (09:58:54.692) - systemUtils.ExecAndPump(<function installAddonBundle at 0x03515438>) (8444):
task 'onInstall' on addon 'wintenApps' failed
Traceback (most recent call last):
  File "addonHandler\__init__.pyc", line 439, in installAddonBundle
  File "addonHandler\__init__.pyc", line 754, in runInstallTask
  File "C:\Users\User\AppData\Roaming\nvda\addons\wintenApps.pendingInstall\installTasks.py", line 14, in onInstall
    raise RuntimeError("Testing add-on install exception handling")
RuntimeError: Testing add-on install exception handling

While not shown in the traceback, this path is called from gui.addonGui.py. This will be important (see technical information).

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
alpha-32397,ad9ed733

#### Windows version:
Windows 11 Version 24H2 preview (build 26100.863, AMD64)

#### Name and version of other software in use when reproducing the issue:
None

#### Other information about your system:
Used as a development workstation

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Reproducible in 2024.1 and 2024.2 beta 3 (see #16702).

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not applicable

### Technical information:
Possible cause: not catching exceptions while retrieving add-on object while calling systemUtils.ExecAndPump during add-on installs. The following code fragments are affected:

* `gui.addonGui.installAddon -> addonObj = systemUtils.ExecAndPump[addonHandler.Addon](addonHandler.installAddonBundle, bundle).funcRes`
* `addonStore.install.installAddon -> addonObj = systemUtils.ExecAndPump[addonHandler.Addon](addonHandler.installAddonBundle, bundle).funcRes`

`systemUtils.execAndPump` runs `addonHandler.installAddonBundle` on a separate thread, with the latter function catching exceptions coming from an add-on's installTasks.onInstall function (two try/except block layers are used). However, right after asking systemUtils.ExecAndPump to run add-on installation routine, NVDA proceeds to remove older add-on releases along with json file(s). This makes the subsequent "add-on install error" dialog dead code (not invoked despite the condition being right to display this dialog).

A possible solution is catching exceptions coming from systemUtils.ExecAndPump through inter-thread communication, with the call to this function housed in its own try/except block so results of possible exceptions will not affect previous add-on instlals. This may also require rethinking the json file handling strategy - previous add-on version json file should not be removed until add-on installation is successful.

Thanks.

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:
Fixes #16704
Fixup of #15967

### Summary of the issue:
In the following code block:

https://github.com/nvaccess/nvda/blob/release-2024.2/source/addonHandler/__init__.py#L432-L454

- If an error is raised from running `addon.runInstallTask("onInstall")`, it is caught, and then re-raised as `AddonError`
- However, this exception is caught again by the `finally` statement `return`ing the add-on
 - [returning from finally suppresses the exception](https://stackoverflow.com/a/40416353/8030743).  This means the callers don't correctly handle the expected exception, as it is suppressed, where it should instead prevent the installation from continuing.
 - We have to return the add-on to perform clean up tasks, but we have to know exceptions were raised to cancel the add-on installation

### Description of user facing changes

Add-on installation failures should fail gracefully

### Description of development approach

Instead of raising exceptions, store them and exit the installation process when a failure occurs.
If exceptions are stored, cancel the installation and perform the require clean-up tasks

### Testing strategy:

**Test external add-on install**
- Download add-on which raises an error on installation [installTasks.nvda-addon.zip](https://github.com/user-attachments/files/16230237/installTasks.nvda-addon.zip)
  - remove `.zip` from the name
- Install the add-on through the Add-on Store (external install button) and through opening an add-on file, 
	- Note on alpha, this installation fails silently as exceptions are ignored
	- Note on this PR, a warning is shown for installation failure

**Test install/update from add-on store**
- Update add-on through add-on store
- Before exiting the store:
  - Go to `%APPDATA%\nvda\addonStore\_dl`
  - replace the downloaded add-on file with the installTasks.nvda-addon. Rename the file to match
- Exit the add-on store
	- Note on alpha, this installation fails silently as exceptions are ignored
	- Note on this PR, a warning is shown for installation failure

### Known issues with pull request:
None

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Enhanced exception handling during addon installation, improving reliability and user feedback.
  - Installation process now provides better logging and error management.

- **Improvements**
  - Improved cleanup handling in the event of installation failure to ensure system stability.
  - Updated error handling to ensure proper cleanup of addon imports and more robust exception management during the installation process.

- **Documentation**
  - Updated function descriptions and parameter documentation for clarity.

- **Refactor**
  - Refactored functions to improve readability and maintainability.
  - Introduced new context manager for better resource management during addon installation.

### CODE REVIEW

**Comment by seanbudd** [comment]
@josephsl - could you test this build? could you also provide a sample add-on to confirm testing with?

**Comment by josephsl** [comment]
Hi,

Tested from the source code copy and things work as expected. To reproduce, use the attached test version of Windows App Essentials build. When the add-on is run with the PR applied, you'll get a message about unsupported Windows release (don't worry about the message itself as we're testing the install failure mechniams).

Thanks.

**Comment by josephsl** [comment]
Be sure to rename from .zip to .nvda-addon extension.

**Comment by seanbudd** [comment]
@josephsl - I think that failed to attach

**Comment by josephsl** [comment]
Hi,

As an alternative, try the following add-on for testing purposes:

https://github.com/josephsl/wintenApps/releases/download/24.06/wintenApps-20240714.0.0.nvda-addon

Thanks.

File: source/addonHandler/__init__.py [hunk 1]
```diff
@@ -427,31 +427,45 @@ def getAvailableAddons(
 def installAddonBundle(bundle: AddonBundle) -> Addon | None:
 	""" Extracts an Addon bundle in to a unique subdirectory of the user addons directory,
 	marking the addon as needing 'install completion' on NVDA restart.
+
+	:param bundle: The add-on bundle to install.
+	The bundle._installExceptions property is modified to store any raised exceptions
+	during the installation process.
+
+	:return: The installed add-on object, or None if the installation failed.
+	Regardless if installation failed or not, the created add-on object should be returned
```
Comment (michaelDCurran) [inline source/addonHandler/__init__.py:436]: Seems this line contradicts the previous line? I.e. returning None if installation fails verses the addon object should ways be returned even if installation fails.

Comment (seanbudd) [inline source/addonHandler/__init__.py:436]:
```suggestion
	:return: The extracted add-on object, or None if the add-on bundle fails to be extracted.
	Regardless if installation failed or not, the created add-on object from the bundle should be returned
```

**Review by michaelDCurran — DISMISSED** [review]

File: source/addonHandler/__init__.py [hunk 2]
```diff
@@ -427,31 +427,45 @@ def getAvailableAddons(
 def installAddonBundle(bundle: AddonBundle) -> Addon | None:
 	""" Extracts an Addon bundle in to a unique subdirectory of the user addons directory,
 	marking the addon as needing 'install completion' on NVDA restart.
+
+	:param bundle: The add-on bundle to install.
+	The bundle._installExceptions property is modified to store any raised exceptions
+	during the installation process.
+
+	:return: The extracted add-on object, or None if the add-on bundle fails to be extracted.
+	Regardless if the add-on installation failed, the created add-on object from the bundle should be returned.
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:436]:
```suggestion
	Regardless if the add-on installation failed, the created add-on object from the bundle should be returned
```

**Pull request merged**